code/game/world.dm

```diff
@@ -88,20 +88,24 @@
 	load_mods()
 	//end-emergency fix
 
+	TgsNew()
+
 	. = ..()
 
 #ifdef UNIT_TEST
 	log_unit_test("Unit Tests Enabled. This will destroy the world when testing is complete.")
 	load_unit_test_changes()
 #endif
 	Master.Initialize(10, FALSE)
+	TgsInitializationComplete()
 
 #undef RECOMMENDED_VERSION
 
 var/world_topic_spam_protect_ip = "0.0.0.0"
 var/world_topic_spam_protect_time = world.timeofday
 
 /world/Topic(T, addr, master, key)
+	TGS_TOPIC
```

Comment:
I feel this is potentially unsafe. 

At a minimum it needs to handle spam protection on input which fails the authentication check, in all cases (that seems to require extra logic from what is shown here). Spam protection seems to be handled via the two vars defined just above on an ad hoc basis.

Moreover, it's probably a good idea to handle the authentication step prior to the processing step, in particular prior to calling expensive stuff like db queries, json_decode on arbitrary input, file i/o, etc. Not doing so leaves you vulnerable to DOS shenanigans even without very high request count.

Would rather someone with better knowledge offered feedback, though.

Reply:
Actually, this entire Topic proc is currently unsafe, as the server writes to logs every single world Topic() call, spam prevention is completely missing from potentially performance-intensive calls such as `status`, and it does not take into account the potential for multiple IPs to be performing an attack simultaneously at any given time.
Aside from that, it's important that TGS is granted some form of bypass for all forms of topic spam prevention, as the watchdog is extremely chatty with the server, especially during bootup. But it'd have to be done in a way that also prevents forged/unauthorized input from abusing it to spam the server.